### ISSUE TITLE
Wrong code for avx2-i64x4

### ISSUE THREAD

**Issue by Vsevolod-Livinskij** [ispc/ispc#1763]
ISPC produces wrong code for avx2-i64x4

>$ ispc --target=avx2-i64x4 -o func.o func.ispc && clang++ driver.cpp func.o && ./a.out
-3978 -1
>$ ispc --target=avx2-i32x8 -o func.o func.ispc && clang++ driver.cpp func.o && ./a.out
-3978 0

Reproducer:
driver.cpp:
```c++
#include <stdio.h>

unsigned long long int var_1 = 9494725476832750138ULL;
bool var_11 = 0;
unsigned long long int arr_253 [2] ;
bool arr_252 [17] [2] ;
short arr_255 [2] ;

void checksum() {
    printf("%d %d\n", arr_255[0], arr_255[1]);
}

extern "C" { void test(); }

int main() {
    for (size_t i_0 = 0; i_0 < 2; ++i_0)
        arr_253 [i_0] = 11458599792036474998ULL;
    test();
    printf("%d %d\n", arr_255[0], arr_255[1]);
}
```

func.ispc
```c++
extern uniform unsigned int64 var_1;
extern uniform bool var_11;
extern uniform unsigned int64 arr_253 [2] ;
extern uniform bool arr_252 [17] [2] ;
extern uniform int16 arr_255 [2] ;

export void test() {
    if (-5174796161658191029LL)
        foreach(i_43 = 0LL...16LL)
            for (uniform bool i_52 = 0; i_52 < 1; i_52 += 1)
                arr_252 [i_43] [i_52] = (uniform unsigned int) ((var_11 ? 195 : 1153317864U) << (var_1 - 9494725476832750132ULL));
    for (uniform bool i_58 = 0; i_58 < 1; i_58 += 1)
        arr_255 [i_58] = arr_253 [i_58];
}
```

ISPC version:  1.14.0dev (build commit 1da899442b7af430 @ 20200520, LLVM 10.0.0)

**Comment by aschrein** [comment]
I think it's aliasing. ISPC assumes bool to be 64 bit for i64x4 and 32 bit for i32x8.
This store:
```c++
arr_252 [i_43] [i_52] = (uniform unsigned int) ((var_11 ? 195 : 1153317864U) << (var_1 - 9494725476832750132ULL));
```
Results in assembly for i32x8 [godbolt](https://godbolt.org/z/J6Zw4f):
```asm
mov     dword ptr [rip + arr_252], eax
```
And this is for i64x4:
```asm
mov     qword ptr [rip + arr_252], rax
```
The law says:
* Representation of bool type in storage was changed from target-specific to one byte per boolean value. So size of varying bool is target width (in bytes), and size of unform bool is one. This definition is compatible with C/C++, hence improves interoperability.

Specifically ` size of varying bool is target width (in bytes)` I'm interpreting this as expected behavior so ISPC's varying bool and C++'s bool are not binary compatible and therefore the attached source is malformed.

But. The strides for stores are quite strange:
```asm
...
mov     qword ptr [rip + arr_252+2], rax
mov     qword ptr [rip + arr_252+4], rax
mov     qword ptr [rip + arr_252+6], rax
...
```
Apparently for stores ISPC assumes sizeof(bool) == 1 but src is 4/8 bytes. This needs further investigation.

**Comment by dbabokin** [comment]
I don't see a problem in the reproducer, it looks like a correct program.
sizeof(uniform bool) == 1, i.e. one byte.
sizeof(varying bool) == 4/8/16... or whatever target width is. It will correspond to C++ bool[4]/bool[8]/bool[16].

This strides also look good, it's declared as ``uniform bool arr_252 [17] [2] ``, so
```asm
mov     qword ptr [rip + arr_252+2], rax // => uniform bool arr_252 [1] [0] ;
mov     qword ptr [rip + arr_252+4], rax // => uniform bool arr_252 [2] [0] ;
mov     qword ptr [rip + arr_252+6], rax // => uniform bool arr_252 [3] [0] ;
```
The difference ``eax`` vs ``rax`` seems to be the only difference and it looks suspicious.

**Comment by dbabokin** [comment]
The problem is the width of bool stores. They are ``i32`` for ``avx2-i32x8`` and ``i64`` for ``avx2-i64x4``. They need to be ``i8``. These are byte stores.

@DeepakRajendrakumaran something is not working as intended after changing definition bool.

**Comment by dbabokin** [comment]
Here's a godbolt [link](https://godbolt.org/z/B3p-4v).

**Comment by DeepakRajendrakumaran** [comment]
Good catch! It seems to be an incorrect check : https://github.com/DeepakRajendrakumaran/ispc/commits/1762_1763

The right version would be

mov     byte ptr [rip + arr_252+2], al // => uniform bool arr_252 [1] [0] ;
mov     byte ptr [rip + arr_252+4], al// => uniform bool arr_252 [2] [0] ;
mov     byte ptr [rip + arr_252+6], al// => uniform bool arr_252 [3] [0] ;

**Comment by dbabokin** [comment]
Fixed by #1766
